### DISCUSSION TITLE
Example of looping wav file

### DISCUSSION THREAD

**Discussion by StacyPointe** [FactoryXCode/MfPack#77] [Q&A] [answered]
Good day,

Can someone share an example of code to loop an audio file, say for example, 3 times when Play is triggered? Below is what I have:

procedure TMainForm.Timer1Timer(Sender: TObject);
var
  position: LongInt;
  FPlayCount : integer;
begin
  Inc(FPlayCount);
    if FPlayCount < 3 then
     begin
     trbPlayBackPosition.Position := 0;
     g_MediaPlayer.Play();
     end;
     end
     else
     begin
       Timer1.Enabled := False;
       Label15.Caption := 'Audio has reached maximum loop limit.';
     end;
end;

The concept is when a process completes, play an audio file.  When the audio file reaches the end, increment FPlayCount and play again (until 3 times has been reached). When FPlayCount = 3, quit playing.

Thank you.

**Comment by FactoryXCode** [comment, marked as answer]
Hi,
Using a timer is usually a bad idea. For instance some media files give an incorrect length.
What you should do, implement this in a callback or when the renderer signals playing has been ended (end of stream).
For example:

If you only want to play audio, then look for example, XAudio2Player Sample 1.

Here you find function TXaudio2Engine.InitializeXAudio2 in unit TXaudio2Engine.

```
  // While the audio is rendering, keep track of the position and samples played.

  TThread.Synchronize(nil,
                      procedure
                      var
                        bReady: Boolean;
                        voiceState: XAUDIO2_VOICE_STATE;

                        begin
                          bReady := False;
                          while not bReady and bPlaying and (pvSourceVoice <> nil) do
                            begin
                              pvSourceVoice.GetState(voiceState,
                                                     0);

                              bReady := (voiceState.BuffersQueued = 0);
                              // Send score. Don't use PostMessage because it set priority above this thread.
                              SendMessage(hwndCaller,
                                          WM_DATA_PROCESSED_NOTIFY,
                                          WPARAM((voiceState.SamplesPlayed div nSamplesPerSecond) * 10000000),
                                          LPARAM(voiceState.SamplesPlayed));
                            end;
                        end);
done:
  if SUCCEEDED(hr) then
    // Signal we are done playing. << On this moment, you can add a counter in the mainform procedure OnAudioEndedEvent.
    SendMessage(hwndCaller,
                WM_DATA_ENDED_NOTIFY,
                WPARAM(1),
                LPARAM(0));

  Result := hr;
```

When the mainform get's the signal the stream has been ended (procedure OnAudioEndedEvent) , just hit Playbutton again, until the number of times plaing has been reached.

Like so:

```
procedure TfrmMain.OnAudioEndedEvent(var AMessage: TMessage);
begin

  if (AMessage.WParam = 1) then
    begin
      StatusBar.SimpleText := Format('Ended playing: %s', [fAudioFileName]);
      butPlayPause.Enabled := False;
      butPlayPause.Caption := 'Play';
      butReplay.Enabled := True;

      // Here we decrease the counter you have set.
      If (MyCounter <> 0) then 
        begin
           Dec(MyCounter);
           butReplayClick(Self); // Start playing again.
       end;
end;
```

Basicly you can do the same with every MfPack sample.

Regards, Tony.

**Reply by StacyPointe** [reply]
Thank you very much.

Stacy
